ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,438 @@
+
+	/* Fill with zeros and convert as many vector elements as needed */
+	size_t n1_vector_count = 0;
+	while (n1_bottom_extension >= BC_VECTOR_SIZE) {
+		n1_vector[n1_vector_count] = 0;
+		n1_bottom_extension -= BC_VECTOR_SIZE;
+		n1_vector_count++;
+	}
+
+	size_t n1_bottom_read_len = BC_VECTOR_SIZE - n1_bottom_extension;
+
+	size_t base;
+	size_t n1_read = 0;
+	if (n1_bottom_read_len < BC_VECTOR_SIZE) {
+		n1_read = MIN(n1_bottom_read_len, n1_readable_len);
+		base = POW_10_LUT[n1_bottom_extension];
+		n1_vector[n1_vector_count] = 0;
+		for (size_t i = 0; i < n1_read; i++) {
+			n1_vector[n1_vector_count] += *n1 * base;
+			base *= BASE;
+			n1--;
+		}
+		n1_vector_count++;
+	}
+
+	/* Bulk convert n1 and n2 to vectors */
+	if (n1_readable_len > n1_read) {
+		bc_convert_to_vector(n1_vector + n1_vector_count, n1, n1_readable_len - n1_read);
+	}
+	bc_convert_to_vector(n2_vector, n2, n2_len);
+
+	/* Do the divide */
+	if (n2_arr_size == 1) {
+		bc_fast_div(n1_vector, n1_arr_size, n2_vector[0], quot_vector, quot_arr_size);
+	} else {
+		bc_standard_div(n1_vector, n1_arr_size, n2_vector, n2_arr_size, n2_len, quot_vector, quot_arr_size);
+	}
+
+	/* Convert to bc_num */
+	char *qptr = (*quot)->n_value;
+	char *qend = qptr + quot_len - 1;
+
+	size_t i;
+	for (i = 0; i < quot_real_arr_size - 1; i++) {
+#if BC_VECTOR_SIZE == 4
+		bc_write_bcd_representation(quot_vector[i], qend - 3);
+		qend -= 4;
+#else
+		bc_write_bcd_representation(quot_vector[i] / 10000, qend - 7);
+		bc_write_bcd_representation(quot_vector[i] % 10000, qend - 3);
+		qend -= 8;
+#endif
+	}
+
+	while (qend >= qptr) {
+		*qend-- = quot_vector[i] % BASE;
+		quot_vector[i] /= BASE;
+	}
+
+	efree(n1_vector);
+}
 
-/* The full division routine. This computes N1 / N2.  It returns
-   true if the division is ok and the result is in QUOT.  The number of
-   digits after the decimal point is SCALE. It returns false if division
-   by zero is tried. The algorithm is found in Knuth Vol 2. p237. */
-bool bc_divide(bc_num n1, bc_num n2, bc_num *quot, int scale)
+bool bc_divide(bc_num n1, bc_num n2, bc_num *quot, size_t scale)
 {
-	bc_num qval;
-	unsigned char *num1, *num2;
-	unsigned char *ptr1, *ptr2, *n2ptr, *qptr;
-	int scale1, val;
-	unsigned int len1, len2, scale2, qdigits, extra, count;
-	unsigned int qdig, qguess, borrow, carry;
-	unsigned char *mval;
-	unsigned int norm;
-	bool zero;
-
-	/* Test for divide by zero. */
+	/* divide by zero */
 	if (bc_is_zero(n2)) {
 		return false;
 	}
 
-	/* Test for divide by 1. If it is we must truncate. */
-	if (n2->n_scale == 0 && n2->n_len == 1 && *n2->n_value == 1) {
-		qval = bc_new_num (n1->n_len, scale);
-		qval->n_sign = (n1->n_sign == n2->n_sign ? PLUS : MINUS);
-		memcpy(qval->n_value, n1->n_value, n1->n_len + MIN(n1->n_scale, scale));
-		bc_free_num (quot);
-		*quot = qval;
+	bc_free_num(quot);
+
+	/* If n1 is zero, the quotient is always zero. */
+	if (bc_is_zero(n1)) {
+		*quot = bc_copy_num(BCG(_zero_));
+		return true;
 	}
```

Comment:
It might make sense to also add a shortcut for when the divisor is `1` in which case the result is just the numerator.

Reply:
Further down, there is a comment that says `/* If n2 is 1 here, return the result of adjusting the decimal point position of n1. */`.

If n2 is 1, it is included in that early return case, but should we return more earlier?

Reply:
I think so, there is no need to do a bunch of processing to handle this case

Reply:
Added
https://github.com/php/php-src/pull/14660/commits/68bcff9eb855af39c3efa19f6cc0e41c8263f5d4